Restore the "Load more" button - no AMP support

WunderBart opened the pull request:
Closes https://github.com/Automattic/newspack-blocks/issues/154

This is a rebased version of https://github.com/Automattic/newspack-blocks/pull/217 with removed commits that introduced the Mustache template. We've decided to continue the approach of reusing existing PHP render function and compile post's HTML into the newly created endpoint response, as planned originally.

obenland committed:
NHA: Bring back More button

First pass, essentially reverting #102

getdave committed:
Moves “more” button outside of CSS Grid

WunderBart committed:
Initialize AMP rendering

WunderBart committed:
Fix typo

getdave committed:
Register basic REST endpoint for Articles Block

getdave committed:
Adjust namespaces. Add HTML to response.

getdave committed:
Tweak endpoint again to remove unwanted addition vendor version

getdave committed:
Amend endpoint data output format

WunderBart committed:
Render HTML string into amp-list template

getdave committed:
Adds next url to response

getdave committed:
Improve method of appending url params

getdave committed:
Make next url relative

Required for AMP to work

WunderBart committed:
Make amp-list src relative
In order to test it locally, you need to either have a localhost or run a HTTPS server (like Flywheel with trusted cert)

obenland committed:
Dynamic rest url

getdave committed:
Restore More button

getdave committed:
Moves article to its own template file via new file include helper

getdave committed:
Move local vars into template

getdave committed:
Fill in some placeholder AMP values

getdave committed:
Move Articles loop into its own template

getdave committed:
Begin extracting attributes into props for templates

getdave committed:
Update REST endpoint to return HTML based on real article query

getdave committed:
Render fallback post list

WunderBart committed:
Use "responsive" layout for the amp-list element
"Elements with "layout=responsive" have no intrinsic size. The size of the element is determined from its container element. To ensure your AMP element displays, you must specify a width and height for the containing element."
https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/control_layout/

getdave committed:
Ensure load more cannot exceed bounds of pagination

getdave committed:
Correct placeholder text

getdave committed:
Mirror article query args used in core theme loop

getdave committed:
Remove “No more articles” indicator as not required

getdave committed:
Refactor conditionals to improve code readability

getdave committed:
DRY up articles query between “frontend” and REST API

getdave committed:
Validate and santize page arg

getdave committed:
Add Block JSON file

getdave committed:
Scaffold out validation callback for attributes

getdave committed:
Always render list even if not in More Button mode

getdave committed:
Fix error attempting validate with method that expects a single arg

WunderBart commented:
I'm having a hard time trying to use the correct layout for the [`amp-list`](https://amp.dev/documentation/examples/components/amp-list/?format=websites) element so the initial content is always rendered at full height - no fixed height, no overflow, etc. The only layout that supports that is `container`, but it's not supported as an initial layout (?). It can be switched to via the `is-layout-container` attribute, but it has to use [`amp-bind`](https://amp.dev/documentation/components/amp-bind/) to do it. AFAIK, value binding can happen only by manual trigger (like a button click), so it won't work on initial load, unfortunately.

I've found an alternative component that seems like a more suitable choice for rendering the latest posts list - the [`amp-live-list`](https://amp.dev/documentation/components/amp-live-list/). Unfortunately, it doesn't support the `Load more` functionality, only pagination 😄 

Our use-case seems to be a mix between what the `amp-list` and the `amp-live-list` are designed for. I will continue the research tomorrow but would love to get any feedback on this. Would switching to `amp-live-list` even make sense if there's no support for the "more" button?

obenland committed:
Move variable setup out of template

Also fixes some phpcs sthings

obenland committed:
Streamline articles-loop

obenland committed:
Properly validate/sanitize attributes

obenland committed:
Reformat block.json

obenland committed:
Update API class

obenland commented:
If we were to make all attributes first-class args to the API, we'd be able to shed the validation/saitization callbacks. We could then just pass the json file as the arguments schema and let the API handle all validation & sanitization. It would also make the `page` argument a little less special and have it just be part of the other query args. Yay/nay?

getdave committed:
Split up render function to allow for alternate render modes.

It may be that we don’t want to force AMP. Therefore we setup to allow the Articles list to be rendered differently depending on some CONSTANT value.

getdave committed:
Uses existing AMP test util to detect AMP support (with dev constant overide)

Add define( 'WP_DEV_AMP_MODE', true ); to force AMP mode for debugging.

getdave committed:
Use Block JSON attr schema in PHP Block registration.

getdave committed:
Update to manually enqueue AMP scripts only if this is not an AMP request

getdave committed:
Update to manually enqueue AMP scripts only if this is not an AMP request

getdave committed:
Creates new standards layout template. Renames AMP version.

getdave committed:
Fix PHP global on $attributes. Add More button

getdave committed:
i18n button text and use correct type

getdave committed:
Rename template inc function to conform to plugin namespace guide

getdave committed:
Update to always renders standards based list.

getdave committed:
Update to always renders standards based list.

WunderBart commented:
After spending some time on the documentation and trying out various configurations for the `amp-list` component, it turns out that it doesn't support dynamic resizing on page load. What that means, is that it requires a fixed height value to be provided for every layout it supports.

I'm considering it a blocker at this point as the content we want to render has variable height, and we'd always want to show it as is, without the `overflow` button to uncover the remaining part.

For more context on this issue, see [this huge thread](https://github.com/ampproject/amphtml/issues/10230) (and [this comment](https://github.com/ampproject/amphtml/issues/10230#issuecomment-376968125)).
_FYI, this [has been supported](https://github.com/ampproject/amphtml/pull/16564) at some point in the past, but it [got rid of due to some issues it caused](https://github.com/ampproject/amphtml/issues/18849)._

Even if we'd be able to use the `amp-list` component (e.g. when the `container` layout gets supported on page-load), there's another thing that would need addressing - The `list` wrapper `div` doesn't accept extra class-names and cannot be opted-out (it's being rendered even if the item's template is empty). This is how it looks: `<div role="list">{items}</div>`. This basically means that we can't use any shared/imported class-names and would have to write/copy additional CSS to keep currently applied styling.

obenland committed:
Condense amp script reggo

obenland committed:
phpcs spacing

obenland committed:
Short array syntax

obenland committed:
Comment updates

obenland committed:
Set propper defaults

obenland committed:
Bring API args to top-level

Let's API handle validation/sanitization based on `block.json`

getdave committed:
Update PHPCS config to remove invalid ref to WordPress.VIP

Note WordPress.VIP standard is no longer included in WPCS standards. Therefore this reference breaks the integration.

getdave committed:
WordPress coding standards auto fixes

getdave committed:
Fix rebase error with Block `name` declared twice.

getdave committed:
Move Category to block.json

getdave committed:
Fix rebase error causing undefined PHP function

Function had moved to a static method on a class rather than a global scope function.

getdave committed:
Fix to only render button if there are more pages to be loaded.

WunderBart committed:
Initialize view.js part for the Load More button

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
 
 import './view.scss';
+
+const loadMoreButton = document.getElementById( 'load-more-button' );
```
Using IDs here could be problematic if we have multiple blocks on the page. I would recommend selecting by a data attribute.

getdave commented on src/blocks/homepage-articles/view.php:
```diff
@@ -92,291 +92,124 @@ function newspack_blocks_render_block_homepage_articles( $attributes ) {
-					$styles = 'min-height: ' . $attributes['minHeight'] . 'vh; padding-top: ' . ( $attributes['minHeight'] / 5 ) . 'vh;';
-				}
+			?>
+			<div id="posts-container" class="<?php echo esc_attr( $classes ); ?>" style="<?php echo esc_attr( $styles ); ?>">
```
Not sure we can use an `id` attr here because it should be unique in the DOM and if we have multiple Blocks on the page if won't be.

See also comment about `getElementByID()`.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
+
+function handleLoadMoreButtonClick() {
+  const loadMoreURL = this.getAttribute( loadMoreButtonURLAttribute );
+  const request = new XMLHttpRequest();
```
We should be able to use `@wordpress/api-fetch` here which wraps the `fetch` API in browsers.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
+      renderPosts( payload.items );
+      loadMoreButton.setAttribute( loadMoreButtonURLAttribute, payload.next );
+    } else {
+      console.log( 'Something went wrong!' );
```
Next step: let's display something to the user if something went wrong. Probably by displaying a human friendly notice and allowing them to try again by re-clicking the button.

WunderBart replied:
Can we use some notice component here?

getdave replied:
Don't think so. Just a paragraph probably. We don't have any components available like this.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
+
+  request.onload = function () {
+    if ( request.status === 200 ) {
+      const payload = JSON.parse( request.responseText );
```
This will throw a SyntaxError exception if it's not valid JSON. We should wrap in `try/catch` and use a error handler to show an notice to the user.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
+};
+
+function renderPosts( items ) {
+  const postsHTML = items.map( item => item.html );
```
I'm confused about this code because `postsHTML` would be an **array** of HTML strings. `insertAdjacentHTML ` expects a string of text as it's second argument.

How is this even working? 😄

WunderBart replied:
Good catch! Yeah, it stringifies it and leaves the `,` between every div. Will update 👍

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
+  request.send();
+};
+
+function renderPosts( items ) {
```
Can we check that when this JS is "built" it is properly "rolled up" inside a closure or something that is protecting the scope of the vars and function declarations from bleeding into the global scope?

WunderBart replied:
It is wrapped as a module, so nothing gets out :)

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -5,3 +5,36 @@
+    if ( request.status === 200 ) {
+      const payload = JSON.parse( request.responseText );
+
+      renderPosts( payload.items );
```
Could we

1. Show a loading indicator?
2. Disable the button from repeated clicks?
3. Only allow a single request at a time (via closure in code). This builds on 2 above.

WunderBart replied:
Ad 1. Should we:
   - use some existing loading component,
   - replace button's text to `Loading...` or
   - hide button and render `Loading...` text (would address no. 2 and _almost_ 3, too)?

getdave replied:
One of the later two.

getdave reviewed (changes requested):
Nice start to the `view.js` for the "Load More" @WunderBart 👍 

A few refinements would be great 🥇

getdave committed:
Comments and docblocks

getdave committed:
Adds qualifier text to “More button” in Block Inspector controls

Adds some additional text to indicate that the “More” button is only available for non-AMP requests. This is because we cannot have JS scripting on AMP and `amp-list` doesn’t yet fully support our use case.

WunderBart committed:
Use data selectors instead of ids

WunderBart committed:
Join mapped posts HTML to render correctly

WunderBart committed:
Use apiFetch instead of XMLHttpRequest

WunderBart committed:
Handle loading state

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,54 @@
+const loadMoreBtn = document.querySelector( '[data-load-more-btn]' );
+const loadMoreBtnURLAttr = 'data-load-more-url';
+const loadMoreLoadingText = document.querySelector( '[data-load-more-loading-text' );
+const postsContainer = document.querySelector( '[data-posts-container]');
```
These selectors should be scoped from the `loadMoreBtn`.

```
const loadMoreLoadingText = loadMoreBtn.querySelector( '[data-load-more-loading-text' );
```

If you don't do this you may select the wrong element.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,54 @@
+}
+
+function handleLoadMoreBtnClick() {
+  const loadMoreURL = this.getAttribute( loadMoreBtnURLAttr );
```
We should probably avoid `this` here if possible. Try and get a ref to the target.

getdave reviewed (changes requested):
Some tweaks.

WunderBart committed:
Make sure fetch is not called multiple times

WunderBart committed:
Add missing bracket

WunderBart committed:
Make button click set correct block scope

WunderBart committed:
Make error handling user friendly; clean up a bit

WunderBart committed:
Hide the button if there's no more posts to load

WunderBart commented:
At this moment, the `view.js` script supports:
  - multiple instances of the `Load more` button by switching block scope,
  - basic, but user-friendly loading and error messages,
  - hiding the `Load more` button if there are no more posts to load (`next` is _falsy_).

WunderBart commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,89 @@
+  isFetching = true;
+
+  apiFetch( { url: loadMoreURL } )
+    .then(( data ) => {
```
@getdave should we validate here if the `data` is a JSON?

getdave replied:
Yes we should. 

Additional: We could also check it conforms roughly to the schema we're expecting, namely objects containing `items` containing a `String`.

WunderBart committed:
More cleaning up

WunderBart committed:
Fix & tweak EOD handling

obenland commented on src/blocks/homepage-articles/view.php:
```diff
@@ -92,291 +92,127 @@ function newspack_blocks_render_block_homepage_articles( $attributes ) {
+				<button type="button" data-load-more-url="<?php echo $articles_rest_url ?>"><?php _e( 'Load more articles' ); ?></button>
+				<p data-load-more-loading-text hidden><?php _e( 'Loading...' ); ?></p>
+				<p data-load-more-error-text hidden><?php _e( 'Something went wrong. Please refresh the page and/or try again.'); ?></p>
+				<p data-load-more-eod-text hidden><?php _e( 'No more posts to load!' ); ?></p>
```
Let's not show that. The lack of a load more button should be indicator enough.

getdave replied:
Agreed 👍

obenland committed:
Tabs > Spaces

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+let isFetching = false;
+let isEndOfData = false;
+
+if ( page ) {
```
I like the event delegation approach here 👍 

Could/should we not simply delegate to the document as that is guaranteed to exist whereas the `id="page"` isn't?

WunderBart replied:
Good point. I assumed `page` is a convention and it will always be there. If there's any possibility that it will not, we should absolutely move to something higher, but I think `body` would be better than the whole document to attach the listener to in this case?

getdave replied:
I've always used `document` on delegation. What do others think @marekhrabe @retrofox @frontdevde @obenland?

marekhrabe replied:
I think it’s smart to use more specific wrapping element for delegation but in wp world, there is no guarantee of a theme markup so I think document might be appropriate

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+}
+
+function handlePageClick( event ) {
+	const isLoadMoreBtnClick = !!event.target.hasAttribute( loadMoreBtnURLAttr );
```
Do we need to check for the existence of `event.target` first?

```js
!! event.target && event.target.hasAttribute( loadMoreBtnURLAttr );
```

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+}
+
+function handleLoadMoreBtnClick( loadMoreBtn ) {
+	if ( !loadMoreBtn || isFetching || isEndOfData ) {
```
My guess is we should have already guarded against `null` values of `loadMoreBtn` but I guess I make sense to be doubly cautious!

getdave replied:
Good to have `isEndOfData` check as well 👍

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+};
+
+function hideEl( el ) {
+	console.log( el.outerHTML );
```
Let's ditch the `console` here.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+				showEl( loadMoreBtn );
+			} else {
+				hideEl( loadMoreBtn );
+				showEl( loadMoreEODText );
```
I think we can remove this as @obenland suggested elsewhere in his review.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+		.then( ( data ) => {
+			renderPosts( postsContainer, data.items );
+
+			if ( data.next ) {
```
Minor: should we _also_ check for the `.length` here? Just in case the endpoint updates and the value is an empty string?

WunderBart replied:
It should be fine - `""` (empty string) is treated as false, too 👍

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+
+	apiFetch( { url: loadMoreURL } )
+		.then( ( data ) => {
+			renderPosts( postsContainer, data.items );
```
We should def validate the data is what we're expecting before we attempt to `renderPosts` here. Probably in a dedicated method.

WunderBart replied:
If the data object isn't valid it will be caught by the promise and the error message will be rendered. I think that would be the correct behaviour in such a situation?

getdave replied:
It will catch invalid JSON, but it won't necessarily match the schema (shape) of the data we're expecting.

I was just thinking we should check for a collection of `items` with an `html` prop which is a `String`.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+			isFetching = false;
+		} )
+		.catch( ( error ) => {
+			console.error( error );
```
Minor: should we only log this when in some sort of dev mode? Trying to recall if WP provides the state of `WP_DEBUG` to JS or not.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+};
+
+function renderPosts( targetEl, items ) {
+	const postsHTML = items.map( item => item.html ).join( '' );
```
Do you think its worth having some sort of guard here against `item.html` being undefined?

getdave reviewed (changes requested):
Looking much better. Nice job! A few tweaks and questions. Let's discuss.

Also we need to be testing and developing against a Post containing multiple instances of this Block.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -1,7 +1,103 @@
+const page = document.getElementById( 'page' );
+const loadMoreBtnURLAttr = 'data-load-more-url';
+let isFetching = false;
+let isEndOfData = false;
```
**Major**: i've just realised this is a global state var. This means that if you have multiple Blocks on the page and your exhaust the posts in one Block (ie: you've loaded all the posts) then this var will Block all the other Blocks from loading more posts.

getdave replied:
Also if one Block is fetching that probably shouldn't stop other Blocks from fetching.

getdave replied:
Let's use closure to contain the state to a per load more button "instance" basis.

WunderBart replied:
Great catch! Will address it in the next commit. Thanks!

getdave reviewed (changes requested):
Found a major bug. Essentially [one Block can stop others from loading more posts](https://cld.wthms.co/woHrMq)!

WunderBart committed:
Make closure for button click handler

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -16,88 +16,93 @@ import './view.scss';
+	let isEndOfData = false;
+
+	btnEl.addEventListener( 'click', function() {
+		if ( isFetching || isEndOfData ) {
```
We should be able to extract the handler `function` definition outside of the event listener. That way we don't redeclare it onclick each time.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -16,88 +16,93 @@ import './view.scss';
 
 function renderPosts( targetEl, items ) {
-	const postsHTML = items.map( item => item.html ).join( '' );
+	if ( !targetEl || !items || items.length === 0 ) {
```
Nit: `!items.length` will have the same effect no?

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -16,88 +16,93 @@ import './view.scss';
-}
+document
+	.querySelectorAll( '[data-load-more-btn]' )
+	.forEach( makeLoadMoreHandler )
```
Missing `;`. Not sure how this getting past the linter...

getdave commented on src/blocks/homepage-articles/view.php:
```diff
@@ -136,10 +136,9 @@ function newspack_blocks_render_block_homepage_articles( $attributes ) {
 
 			if ( ! Newspack_Blocks::is_amp() && $has_more_pages ) : ?>
-				<button type="button" data-load-more-url="<?php echo $articles_rest_url ?>"><?php _e( 'Load more articles' ); ?></button>
+				<button type="button" data-load-more-btn data-load-more-url="<?php echo $articles_rest_url ?>"><?php _e( 'Load more articles' ); ?></button>
```
`data-load-more-url="<?php echo $articles_rest_url ?>` needs escaping using the appropriate `esc_*` function.

getdave commented on src/blocks/homepage-articles/view.js:
```diff
@@ -16,88 +16,93 @@ import './view.scss';
+				isFetching = false;
+			} )
+			.catch( ( error ) => {
+				// console.error( error );
```
Can we remove this comment?

getdave reviewed (changes requested):
I've tested this and it now works functionality wise. Great 👍 

Some code points to tidy up.

WunderBart committed:
Format with prettier; minor cleanup

getdave committed:
Slight tweak to reorganise the closure.

WunderBart committed:
Format with prettier; minor cleanup

WunderBart committed:
Escape load more URL

getdave commented:
Just noticed a bug with https://github.com/Automattic/newspack-blocks/blob/a1a11c41ba2018ac81860192c88975390bc571f6/src/blocks/homepage-articles/edit.js#L146.

`post.title.rendered` may or may not exist. We need to test.

getdave commented:
## Status of this PR

* May require splitting into two PRs: one for refactor and one for the REST API work. Possibly overkill? 
* Fix some minor code hygiene points.

obenland commented on src/blocks/homepage-articles/edit.js:
```diff
@@ -267,6 +269,14 @@ class Edit extends Component {
+							label={ __( 'Show "More" Button' ) }
+							checked={ moreButton }
+							onChange={ () => setAttributes( { moreButton: ! moreButton } ) }
+							help={ __('Only available for non-AMP requests.', 'newspack-blocks') }
```
I worry this help text would be more confusing than helpful to the average user

WunderBart replied:
I agree. I guess making the text less confusing is one option, but wouldn't it be even better if we just hide the whole toggle for non-AMP requests?

jeffersonrabb replied:
This text appears in the editor and the concept of being an AMP request only really applies to the view. It's conceivable the text could be shown only if the AMP plugin is activated but that seems complex and error-prone.

WunderBart replied:
Thanks, @jeffersonrabb - I've totally missed that fact! Keeping the text to make editors aware makes total sense.

WunderBart committed:
Fix block namespace prefix
Fixes "Notice: WP_Block_Type_Registry::register was called incorrectly. Block type names must contain a namespace prefix. Example: my-plugin/my-custom-block-type"

WunderBart committed:
Add endpoint data validation & more comments

WunderBart committed:
Clean up buildLoadMoreHandler closure; more comments

jeffersonrabb commented:
Closing, as this work will continue in https://github.com/Automattic/newspack-blocks/pull/259.
